pkg/builder/kaniko/kaniko.go

```diff
@@ -0,0 +1,248 @@
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error listing Pods: %w", err)
+	}
+
+	if len(podList.Items) == 0 {
+		return nil, fmt.Errorf("no Pods found for the Job: %s", job.Name)
+	}
+
+	return &podList.Items[0], nil
+}
+
+func (k *Kaniko) getContainerLogs(ctx context.Context, pod *v1.Pod) (string, error) {
+	if len(pod.Spec.Containers) == 0 {
+		return "", fmt.Errorf("no containers found in Pod: %s", pod.Name)
+	}
+
+	containerName := pod.Spec.Containers[0].Name
+
+	logOptions := v1.PodLogOptions{
+		Container: containerName,
+	}
+
+	req := k.k8sClientset.CoreV1().Pods(k.k8sNamespace).GetLogs(pod.Name, &logOptions)
+	logs, err := req.DoRaw(ctx)
+	if err != nil {
+		return "", fmt.Errorf("error retrieving container logs: %w", err)
+	}
+
+	return string(logs), nil
+}
+
+func (k *Kaniko) cleanup(ctx context.Context, job *batchv1.Job) error {
+	err := k.k8sClientset.BatchV1().Jobs(k.k8sNamespace).
+		Delete(ctx, job.Name, metav1.DeleteOptions{
+			PropagationPolicy: &[]metav1.DeletionPropagation{metav1.DeletePropagationBackground}[0],
+		})
+	if err != nil {
+		return fmt.Errorf("error deleting Job: %w", err)
+	}
+
+	// Delete the associated Pods
+	err = k.k8sClientset.CoreV1().Pods(k.k8sNamespace).
+		DeleteCollection(ctx, metav1.DeleteOptions{}, metav1.ListOptions{
+			LabelSelector: fmt.Sprintf("job-name=%s", job.Name),
+		})
+	if err != nil {
+		return fmt.Errorf("error deleting Pods: %w", err)
+	}
+
+	return nil
+}
+
+func GetDefaultCacheOptions(buildContext string) (*builder.CacheOptions, error) {
+	if buildContext == "" {
+		return nil, fmt.Errorf("build context cannot be empty")
+	}
+	hash := sha256.New()
+	_, err := hash.Write([]byte(buildContext))
+	if err != nil {
+		return nil, err
+	}
+	hashStr := hex.EncodeToString(hash.Sum(nil))
+
+	return &builder.CacheOptions{
+		Enabled: true,
+		Dir:     "",
+		Repo:    fmt.Sprintf("ttl.sh/%s:24h", hashStr),
+	}, nil
+}
+
+func prepareJob(b *builder.BuilderOptions) (*batchv1.Job, error) {
+	jobName, err := generateK8sName(kanikoJobNamePrefix)
+	if err != nil {
+		return nil, fmt.Errorf("error generating Job name: %w", err)
+	}
+
+	oneInt32 := int32(1)
+	job := &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: jobName,
+		},
+		Spec: batchv1.JobSpec{
+			Parallelism: &oneInt32, // Set parallelism to 1 to ensure only one Pod
+			Template: v1.PodTemplateSpec{
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{
+						{
+							Name:  kanikoContainerName,
+							Image: kanikoImage, // debug has a shell
+							Args: []string{
+								`--context=` + b.BuildContext,
+								// TODO: see if we need it or not
+								// --git gitoptions    Branch to clone if build context is a git repository (default branch=,single-branch=false,recurse-submodules=false)
+
+								// TODO: we might need to add some options to get the auth token for the registry
+								"--destination=" + b.Destination,
+							},
+						},
+					},
+					RestartPolicy: "IfFailed",
```

Comment:
Not sure if the `RestartPolicy` is set correctly. Right now if there is something wrong with the `Dockerfile` it fails and keeps trying so it is assumed that there is nothing wrong with the target application build process.
We might want to change this. Please share your thoughts on it

Reply:
should this be OnFailure? If you want it to die, it should be `Never` I believe, what is desired behavior?

Reply:
I think we can leave it to try a couple of times then fail completely.

Reply:
@jrmanes it seems like the default value for `BackoffLimit` is `6` can I just use that instead of introducing the index thing?